### PR TITLE
Add more settings tests - completely cover project\sln level support across products

### LINKED ISSUE #8927: Add more settings tests - completely cover project\sln level support across products
Add more settings tests - completely cover project\sln level support across products

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/8927
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: While reviewing @zkat's work, I realized that not everything that we do with the configs is properly tested.
This PR changes that :)

## Testing/Validation

Tests Added: Yes  
Reason for not adding tests:  
Validation:

### CODE REVIEW

**Review by zkat — APPROVED** [review]

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 1]
```diff
@@ -155,13 +155,111 @@ public void RestoreNetCore_IfAllProjectsAreWithoutRestoreTargetsVerifySuccess()
...
+                var packageId = "packageA";
+
+                await SimpleTestPackageUtility.CreatePackagesAsync(
+                    pathContext.PackageSource,
+                    new SimpleTestPackageContext()
+                    {
+                        Id = packageId,
+                        Version = "1.0.0"
+                    }
+                    );
```
Comment (dtivel) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: Move up to end of previous line.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 2]
```diff
@@ -155,13 +155,111 @@ public void RestoreNetCore_IfAllProjectsAreWithoutRestoreTargetsVerifySuccess()
...
+                    var source = Path.Combine(pathContext.WorkingDirectory, $"source{number}");
+
+                    await SimpleTestPackageUtility.CreatePackagesAsync(
+                        source,
+                        new SimpleTestPackageContext()
+                        {
+                            Id = packageId,
+                            Version = $"{number}.0.0"
+                        }
+                        );
```
Comment (dtivel) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: Move up to end of previous line.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 3]
```diff
@@ -155,13 +155,111 @@ public void RestoreNetCore_IfAllProjectsAreWithoutRestoreTargetsVerifySuccess()
...
+                    File.WriteAllText(configPath, doc.ToString());
+                }
+
+                solution.Create(pathContext.SolutionRoot);
+
+                // Act
+                var r = Util.Restore(pathContext, pathContext.SolutionRoot, expectedExitCode: 0);
+
+                // Assert
+                Assert.True(projects.Count > 0);
```
Comment (dtivel) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]:
```C#
Assert.NotEmpty(projects);
```

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 4]
```diff
@@ -155,13 +155,111 @@ public void RestoreNetCore_IfAllProjectsAreWithoutRestoreTargetsVerifySuccess()
...
+                    sourceEntry.Add(new XAttribute(XName.Get("value"), source));
+                    packageSources.Add(sourceEntry);
+
+                    File.WriteAllText(configPath, doc.ToString());
+                }
+
+                solution.Create(pathContext.SolutionRoot);
+
+                // Act
+                var r = Util.Restore(pathContext, pathContext.SolutionRoot, expectedExitCode: 0);
```
Comment (dtivel) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:243]: Should you validate `r` (=> `result`)?  For example:
```C#
Assert.True(result.Success);
```

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 5]
```diff
@@ -155,13 +155,111 @@ public void RestoreNetCore_IfAllProjectsAreWithoutRestoreTargetsVerifySuccess()
...
+
+                // Assert
+                Assert.True(projects.Count > 0);
+
+                foreach (var number in projects.Keys)
+                {
+                    projects[number].AssetsFile.Libraries.Select(e => e.Name).Should().Contain(packageId);
+                    projects[number].AssetsFile.Libraries.Single(e => e.Name.Equals("packageA")).Version.ToString().Should().Be("1.0.0");
```
Comment (dtivel) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: `"packageA"` => `packageId`
`"1.0.0"` => `packageVersion`

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 6]
```diff
@@ -155,13 +155,111 @@ public void RestoreNetCore_IfAllProjectsAreWithoutRestoreTargetsVerifySuccess()
...
+        /// </summary>
+        [Fact]
+        public async Task RestoreNetCore_WithNuGetExe_WhenRestoringASolution_VerifyPerProjectConfigSourcesAreNotUsed()
+        {
+            // Arrange
+            using (var pathContext = new SimpleTestPathContext())
+            {
+                var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);
+                var projects = new Dictionary<string, SimpleTestProjectContext>();
+                var packageId = "packageA";
```
Comment (dtivel) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: Consider:
```C#
const string packageId = "packageA";
```

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 7]
```diff
@@ -155,13 +155,111 @@ public void RestoreNetCore_IfAllProjectsAreWithoutRestoreTargetsVerifySuccess()
...
+
```
Comment (dtivel) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:172]:
```C#
const string packageVersion = "1.0.0";
```

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 8]
```diff
@@ -155,13 +155,111 @@ public void RestoreNetCore_IfAllProjectsAreWithoutRestoreTargetsVerifySuccess()
...
+                await SimpleTestPackageUtility.CreatePackagesAsync(
+                    pathContext.PackageSource,
+                    new SimpleTestPackageContext()
+                    {
+                        Id = packageId,
+                        Version = "1.0.0"
```
Comment (dtivel) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: `packageVersion`

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 9]
```diff
@@ -155,13 +155,111 @@ public void RestoreNetCore_IfAllProjectsAreWithoutRestoreTargetsVerifySuccess()
...
+                    }
+                    );
+
+                foreach (var number in new[] { "2", "3", "4"})
```
Comment (dtivel) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: Do you need to use 3 projects?  Wouldn't 1 suffice?

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetRestoreTests.cs [hunk 10]
```diff
@@ -354,8 +356,211 @@ public async Task DotnetRestore_LockedMode_NewProjectOutOfBox()
...
+        {
+            // Arrange
+            using (var pathContext = new SimpleTestPathContext())
+            {
+                var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);
+                var projects = new Dictionary<string, SimpleTestProjectContext>();
+                var sources = new List<string>();
+                var projFramework = FrameworkConstants.CommonFrameworks.Net462;
+
+                foreach (var letter in new[] { "A", "B", "C", "D" })
```
Comment (dtivel) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetRestoreTests.cs]: This creates 4 projects while the XML doc comments says 3.  Wouldn't 1 suffice?

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetRestoreTests.cs [hunk 11]
```diff
@@ -354,8 +356,211 @@ public async Task DotnetRestore_LockedMode_NewProjectOutOfBox()
...
+        public async Task DotnetRestore_VerifyPerProjectConfigSourcesAreUsedForChildProjectsWithSolutionAsync()
+        {
+            // Arrange
+            using (var pathContext = new SimpleTestPathContext())
+            {
+                var projects = new Dictionary<string, SimpleTestProjectContext>();
+                var sources = new List<string>();
+                var projFramework = FrameworkConstants.CommonFrameworks.Net462;
+
+                foreach (var letter in new[] { "A", "B", "C", "D" })
```
Comment (dtivel) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetRestoreTests.cs]: 4 vs 3.  Again, why not 1?

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/GetRestoreSettingTaskTests.cs [hunk 12]
```diff
@@ -512,6 +512,134 @@ public void TestConfigFileProbingDirectoryWithSettingsLoadingContext()
...
+                var configFile = Path.Combine(testDir, Settings.DefaultSettingsFileName);
+                File.WriteAllText(configFile, RootConfig);
+
+                // Act
+                var result = task.Execute();
+
+                // Assert
+                result.Should().BeTrue();
+                task.OutputSources.ShouldBeEquivalentTo(new[] { "https://api.nuget.org/v3/index.json" });
+                task.OutputFallbackFolders.ShouldBeEquivalentTo(new string[] {});
```
Comment (dtivel) [inline test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/GetRestoreSettingTaskTests.cs]: `Array.Empty<string>()`?  Same for subsequent tests.

**Review by dtivel — APPROVED** [review]

**Pull request merged**